text/MateuszPabian/main.tex

```diff
@@ -0,0 +1,186 @@
+\State $\gamma := S.acm.root$ \Comment{równoważnie $\gamma := S.acm.root$}
+
+\While{ $True$ } \Comment{faza skanowania, zachowany niezmiennik $\Gamma$}
+    \While {$\gamma{.depth} \geq \frac{m}{2}$} \Comment{szukamy odpowiednio długiego prefiksu wzorca}
+        \If{$\gamma$ is final state}
+            \State \textbf{yield} $(w, i)$ dla każdego $w \in \gamma{.out}$
+        \EndIf
+        \State $i = i + 1$
+        \If{$i \geq n$}
+            \State \textbf{return}
+        \EndIf
+        \State $\gamma = NEXT1(\gamma, text[i])$  \Comment{wykonujemy krok w AC}
+    \EndWhile
+    \State $crit\_pos := i - \gamma{.depth} + 1$ \Comment{pierwsza pozycja, której nie wykluczyliśmy}
+    \State $SHIFT := m - \gamma{.depth}$ 
+    \State $i = i + SHIFT$ \Comment{przeskakujemy fragment tekstu}
+    \If{$i \geq n$}
+        \State \textbf{return}
+    \EndIf
+    \State $\gamma = NEXT2(crit\_pos, i)$ \Comment{szukamy najdłuższego sufiksu w DAWG}
+\EndWhile
+\EndProcedure
+\end{algorithmic}
+\end{algorithm}
+
+Wynikiem tej częsci jest lista par postaci $(w, i)$, gdzie $w$ to któryś z szukanych wzorców, natomiast $i$ to pozycja w tekście, na której został znaleziony wzorzec $w$.
+
+\section{Złożoność obliczeniowa algorytmu}
+
+\begin{LEM}
+Zakładjąc zbudowaną strukturę $AC$ oraz $DAWG$:
+\begin{itemize}
+    \item $NEXT1(\gamma, a)$ działa w czasie $O(1)$. \\
+        Wykonanie funkcji $NEXT1$ sprowadza się do wykonania jednego kroku w automacie AC.\qed
+    \item $NEXT2(j, i)$ działa w czasie $O(min(i-j, i - crit\_pos))$. \\
+        Wykonanie funkcji $NEXT2$ wymaga przeskanowania tekstu od pozycji i do j idąc od prawej. Równoczesne symulowanie DAWG wykona taką samą liczbę kroków.\qed
+\end{itemize}
+\end{LEM}
+
+\begin{TW}
+Algorytm w sumie porównuje co najwyżej $2n$ znaków.
+\end{TW}
+
+\begin{PROF}
+Zauważmy, że znak na danej pozycji może być przetworzony przez $PROCESS1$ co najwyżej raz. Jeśli tak się stało, to następnie zostanie tylko raz przetworzony przez $PROCESS2$ i już nigdy ponownie do niego nie wrócimy. Jeśli natomiast $PROCESS2$ odwiedzi dany znak dwa razy, mamy wtedy pewność, że dany znak nie został przeczytany przez $PROCESS1$ w ogóle. \qed
+\end{PROF}
+
+Niech $\sigma$ to prawdopodobieństwo jednostajengo wystapienia symbolu w tekście. Wtedy:
+
+\begin{LEM}
+$P(\Delta(i) > (3k + 1)*log_{\sigma}m) \leq \frac{1}{m^{k+1}}$, gdzie $P$ to prawdopodobieństwo.
+\end{LEM}
+
+\begin{TW}
+Zakładając $M \leq m^k$ algorytm wykonuje $O(\frac{n}{m} * log_{\sigma} m)$ porównań.
+\end{TW}
+
+\begin{PROF}
+Podczas działania algorytmu wykonywanych jest co najwyżej $O(\frac{n}{m})$ przesunięć.
+Wystarczy jeszcze udowodnić, że średnio pierwsza i druga faza algorytmu odwiedza zaledwie logarytmiczną liczbę znaków. Niech $K = (2k + 1) * log _{\sigma} m$. $PROCESS1$ zakończy się z dużym prawdopodobieństwem po K krokach. Jeśli nie, zrobi on $m^k$ kroków z prawdopodobieństwem $1/m^{k+1}$. Dostajemy więc średnią złożoność $O(K)$ co jest logarytmiczne.
```

Comment:
"zakończy się z dużym prawdopodobieństwem po K krokach" - could you explain it further why it's the case. I know it's a simple application of the lemma above, but still it'd be nice to have it referenced here.